File: entregas/src/main/java/br/com/bota/event/RecebimentoPacoteEntregaKafkaListener.java
package br.com.bota.event;

import br.com.bota.dto.EntregaDTO;
import br.com.bota.service.EntregaService;
import io.micronaut.configuration.kafka.annotation.KafkaKey;
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.tracing.annotation.ContinueSpan;
import io.micronaut.tracing.annotation.SpanTag;
import lombok.extern.slf4j.Slf4j;

import javax.inject.Inject;

@Slf4j
@KafkaListener
public class RecebimentoPacoteEntregaKafkaListener {
    @Inject
    private EntregaService service;

    @Topic("pedido_entrega")
    @ContinueSpan
    public void receive(@SpanTag("entrega.recebida") @KafkaKey long key, EntregaDTO entregaDTO) {
        log.info("Evento recebido: {}-{}", key, entregaDTO);
        service.criarEntrega(EntregaDTO.toDocument(entregaDTO));
    }
}

File: loja-kotlin/kentregas/src/main/kotlin/br/com/bota/event/RecebimentoPacoteEntregaKafkaListener.kt
package br.com.bota.event

import br.com.bota.converter.EntregaConverter
import br.com.bota.lojalib.dto.EntregaDTO
import br.com.bota.service.EntregaService
import io.micronaut.configuration.kafka.annotation.KafkaKey
import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.tracing.annotation.ContinueSpan
import io.micronaut.tracing.annotation.SpanTag
import jakarta.inject.Inject
import org.slf4j.LoggerFactory

@KafkaListener
open class RecebimentoPacoteEntregaKafkaListener @Inject constructor(private val entregaService: EntregaService) {
    val logger = LoggerFactory.getLogger(RecebimentoPacoteEntregaKafkaListener::class.java)

    @Topic("pedido_entrega")
    @ContinueSpan
    open fun receive(@SpanTag("entrega.recebida") @KafkaKey key: Long, dto: EntregaDTO) {
        logger.info("Evento recebido: {}-{}", key, dto)

        entregaService.criarEntrega(EntregaConverter.toEntrega(dto))
    }
}